README
This project is an implementation of commit / post-recieve hooks for various
VCS systems, that send a notification to an AMQP server containing payload
that follows the github payload spec :

https://github.com/github/github-services/blob/master/docs/github_payload

Unit test cases

1. Check that when a commit is done a new payload is sent both by SVN and GIT
   hooks
    - mimic commit retrievial
    - mimic payload sending (implement TestPublisher fake class)
    - make sure that JSON data follows the specs from the URL above

2. Check publishers
    - mimic pika with Mox
    - make sure that the whole code is covered with tests

3. Check base consumer
    - mimic pika with Mox
    - make sure that the whole code is covered with tests


debian/changelog
ether (0.7.9) unstable; urgency=low

  * adapted github shovel to the new setup

 -- Mikhail Sobolev <mikhail.sobolev@nokia.com>  Tue, 29 Mar 2011 17:02:47 +0300

ether (0.7.8) unstable; urgency=low

  * Made payload structure equal for svn and git publishers

 -- Ed Bartosh <Ed Bartosh <eduard.bartosh@nokia.com>>  Wed, 23 Mar 2011 16:02:09 +0200

ether (0.7.7) unstable; urgency=low

  * re-vamped test cases

 -- Mikhail Sobolev <mikhail.sobolev@nokia.com>  Wed, 23 Mar 2011 10:54:59 +0200

ether (0.7.6) unstable; urgency=low

  * Fixed testcases
  * Implemented getting repo url for git hooks

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Tue, 22 Mar 2011 15:44:31 +0200

ether (0.7.5) unstable; urgency=low

  * unified way pika.SelectConnection is used

 -- Anton Berezin <anton.berezin@nokia.com>  Tue, 22 Mar 2011 11:32:42 +0200

ether (0.7.4) unstable; urgency=low

  * Fixed variable name

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Tue, 22 Mar 2011 10:26:13 +0200

ether (0.7.3) unstable; urgency=low

  * SvnHookAPI: fixed basedir path

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Tue, 22 Mar 2011 10:05:25 +0200

ether (0.7.2) unstable; urgency=low

  * added python-coverage to the list of build dependencies
  * hookman API: Fixed paths and plugin category

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Mon, 21 Mar 2011 16:18:32 +0200

ether (0.7.1) unstable; urgency=low

  * fixed testcases
  * Code beautification
  * Added 'heartbeat' attribute to configuration structure
  * Adopted ether-post-receive to refactored API
  * Fixed loading of json

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Fri, 18 Mar 2011 16:22:18 +0200

ether (0.7.0) unstable; urgency=low

  * Added hookman package
  * Fixed package naming scheme

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Tue, 15 Mar 2011 17:18:42 +0200

ether (0.6.1) unstable; urgency=low

  * updated versioned dependencies

 -- Mikhail Sobolev <mikhail.sobolev@nokia.com>  Tue, 15 Mar 2011 14:10:18 +0200

ether (0.6.0) unstable; urgency=low

  * Reimplemented configuration handling
  * Refactored publisher and consumer APIs
  * Simplified directory structure
  * Got rid of Payload class
  * excluded examples from the package
  * Fixed a minor bug.
  * Replaced custom runner with nose.

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Fri, 11 Mar 2011 16:56:12 +0200

ether (0.5.0) unstable; urgency=low

  [Anton Berezin]
  * Added unittests

  [Ed Bartosh]
  * Simplified architecture

 -- Anton Berezin <anton.berezin@nokia.com>  Thu, 10 Mar 2011 18:40:07 +0200

ether (0.4.0) unstable; urgency=low

  * Added dependencies to debian package
  * Code butification.
  * Pylinted
  * Implemented repo url passing for svn hooks

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Tue, 08 Mar 2011 11:52:27 +0200

ether (0.3.1) unstable; urgency=low

  * Repackaged configs to be in the same package as the hooks.

 -- Anton Berezin <anton.berezin@nokia.com>  Mon, 07 Mar 2011 17:15:28 +0200

ether (0.3.0) unstable; urgency=low

  * Complete rewrite of the git hook
  * added test_publisher example
  * fixed python packaging
  * Added test_consumer example
  * Moved test consumer out of API module
  * Reimplemented config handling
  * Added routing_key as a parameter to process_payload method
  * svn hook: Fixed payload structure
  * Added new abstract method process_payload
  * Basic utilility class and script that uses it.
  * Use setproctitle module if available
  * Renamed SvnHooks to SvnHook in the script.
  * Refactored SVN hook.
  * Clean exit on ctrl-c
  * pylint fixes
  * Fixed github shovel script
  * Github relay
  * Baseclass should not print
  * Add command to build documentation coverage.
  * Simplified GitHook logic
  * Documented the code
  * Made GitHook code easy to test
  * parameter re-formatting
  * ignore more build artefacts
  * finalized project rename in setup.py

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Mon, 07 Mar 2011 11:13:02 +0200

ether (0.2.0) unstable; urgency=low

  * Renamed project

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Tue, 01 Mar 2011 16:37:36 +0200

ether (0.1.1) unstable; urgency=low
  
  * Declared queue on listener side
  * Declared exchange on sender side
  * Removed trailing whitespaces
  * Adopted listeners and senders to new config format
  * Made separate configurations for consumer and publisher

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Tue, 01 Mar 2011 13:20:51 +0200

ether (0.1.0) unstable; urgency=low

  * Designed project modules hierarchy
  * Implemented and tested working prototype
  * Setup documentation infrastructure
  * Prepared code style and part of code documentation

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Fri, 25 Feb 2011 21:15:40 +0200

ether (0.0.0) unstable; urgency=low

  * Initial release

 -- Ed Bartosh <eduard.bartosh@nokia.com>  Thu, 24 Feb 2011 17:07:34 +0200



ether/fixtures.py
def mock_get_revlist(old, new):
    return """commit e3b0f9c9a0945b0efa9a9295b24cf6108a6f97bb
Author: Anton Berezin <anton.berezin@nokia.com>
Date:   Wed Mar 2 14:13:14 2011 +0200

    C1

commit ad45a931e70e62bde1f3f03e34c27613c85d07d8
Author: Anton Berezin <anton.berezin@nokia.com>
Date:   Wed Mar 2 14:11:14 2011 +0200

    Took Ed's comments converning not modifying the code to ease the tests.

commit 0161cca7fec8a4b5c9ca9fccf7a393a32222da99
Author: Anton Berezin <anton.berezin@nokia.com>
Date:   Wed Mar 2 11:58:45 2011 +0200

     * Simplified GitHook logic
     * Documented the code
     * Made GitHook code easy to test

"""

git_postreceive_expected_output = {
  'commits': [
  { 'timestamp': '2011-03-02T14:13:14+0200',
    'message': 'C1',
    'id': 'e3b0f9c9a0945b0efa9a9295b24cf6108a6f97bb',
    'author': {'name': 'Anton Berezin', 'email': 'anton.berezin@nokia.com'}},
  { 'timestamp': '2011-03-02T14:11:14+0200',
    'message': "Took Ed's comments converning not modifying the code to ease the tests.",
    'id': 'ad45a931e70e62bde1f3f03e34c27613c85d07d8',
    'author': {'name': 'Anton Berezin', 'email': 'anton.berezin@nokia.com'}},
  { 'timestamp': '2011-03-02T11:58:45+0200',
    'message': '* Simplified GitHook logic\n     * Documented the code\n     * Made GitHook code easy to test',
    'id': '0161cca7fec8a4b5c9ca9fccf7a393a32222da99',
    'author': {'name': 'Anton Berezin', 'email': 'anton.berezin@nokia.com'}}
  ],
  'after': 'e3b0f9c9a0945b0efa9a9295b24cf6108a6f97bb',
  'ref': 'refs/head/master',
  'repository': {
      'url': '',
      'owner': {'email': '', 'name': ''},
      'name': '',
      'description': ''
      },
  'before': '0161cca7fec8a4b5c9ca9fccf7a393a32222da99'
  }

def mock_svnlook(what, repos, rev):
    data = {
        "changed": "U   test\n",
        "log": "commit 20\n",
        "author": "ed\n",
        "date": "2011-02-10 11:33:32 +0200 (Thu, 10 Feb 2011)\n"
    }
    return [data[what]]

svn_postcommit_expected_output = {
  'commits': [{
    'added': [],
    'author': {'name': ['ed\n'], 'email': ''},
    'timestamp': ['2011-02-10 11:33:32 +0200 (Thu, 10 Feb 2011)\n'],
    'modified': ['test'],
    'message': ['commit 20\n'],
    'removed': [],
    'id': '/var/svn/repo'}]}
